### ISSUE TITLE
Blocking of Write* and other methods methods

### ISSUE THREAD

**Issue by rzikm** [rzikm/dotnet-runtime#3]
Currently it is inclear to me how the Write* methods on QuicStream should behave and what the caller
should expect. In general, it would be nice to return from the async method after the data has been
sent, allowing the user to recycle the buffer to provide more data. However since the data may need
to be retransmitted due to packet loss, we may either:

- need to wait until corresponding ACK arrives from the server
- buffer (perhaps only sum suffix part) of the input data internally and return early, transmitting
  the buffered part in the background. This might be more efficient, but also makes the
  implementation more complex.
  
Both the above behaviors will of course need to block either way, since the Write calls need to
respect flow control limits on the particular stream.

**Comment by rzikm** [comment]
Maybe we should strive for same behavior as the sockets do, the [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.send?view=netcore-3.1)  states that:

> A successful completion of the Send method means that the underlying system has had room to buffer your data for a network send.

This implies blocking when control flow limit for the connection/stream cannot be reserved for the data to be sent. Other parts of the documentation specify how safely close connection only after all data have been safely received:

> When using a connection-oriented Socket, always call the Shutdown method before closing the Socket. This ensures that all data is sent and received on the connected socket before it is closed.

Another mystery solved, last one is the Flush method, which I still am not sure what exactly it should do. It is probably intended to make sure any buffered data makes it to the wire even if no more data is sent. It might be also possible that it should work in similar way to Shutdown, but without terminating the stream.

**Comment by rzikm** [comment]
From discussion with MS dev team members: The original API has been designed to block until the provided buffers are no longer needed, be it that it has been sent or it has been buffered internally. This approach has already been implemented, thus this issue is resolved.
